proposals/csharp-9.0/unconstrained-type-parameter-annotations.md

```diff
@@ -0,0 +1,78 @@
+﻿# Unconstrained type parameter annotations
+
+## Summary
+
+Allow nullable annotations for type parameters that are not constrained to value types or reference types: `T?`.
+```C#
+static T? FirstOrDefault<T>(this IEnumerable<T> collection) { ... }
+```
+
+## `?` annotation
+
+In C#8, `?` annotations could only be applied to type parameters that were explicitly constrained to value types or reference types.
+In C#9, `?` annotations can be applied to any type parameter, regardless of constraints.
+
+Unless a type parameter is explicitly constrained to value types, annotations can only be applied within a `#nullable enable` context.
+
+If a type parameter `T` is substituted with a reference type, then `T?` represents a nullable instance of that reference type.
+```C#
+var s1 = new string[0].FirstOrDefault();  // string? s1
+var s2 = new string?[0].FirstOrDefault(); // string? s2
+```
+
+If `T` is substituted with a value type, then `T?` represents an instance of `T`.
+```C#
+var i1 = new int[0].FirstOrDefault();  // int i1
+var i2 = new int?[0].FirstOrDefault(); // int? i2
+```
+
+If `T` is substituted with an annotated type `U?`, then `T?` represents the annotated type `U?` rather than `U??`.
+```C#
+var u1 = new U[0].FirstOrDefault();  // U? u1
+var u2 = new U?[0].FirstOrDefault(); // U? u2
+```
+
+For return values, `T?` is equivalent to `[MaybeNull]T`.
+For argument values, `T?` is equivalent to `[AllowNull]T`.
```

Comment:
Should specify that this applies to inheritance scenarios. Essentially that they can be interchanged there. Really important for code versioning across assembly boundaries.